app/recorder.go

```diff
@@ -0,0 +1,69 @@
+// Copyright 2021 Northern.tech AS
+//
+//    Licensed under the Apache License, Version 2.0 (the "License");
+//    you may not use this file except in compliance with the License.
+//    You may obtain a copy of the License at
+//
+//        http://www.apache.org/licenses/LICENSE-2.0
+//
+//    Unless required by applicable law or agreed to in writing, software
+//    distributed under the License is distributed on an "AS IS" BASIS,
+//    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//    See the License for the specific language governing permissions and
+//    limitations under the License.
+
+package app
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"github.com/mendersoftware/deviceconnect/store"
+)
+
+type Recorder struct {
+	sessionID string
+	store     store.DataStore
+	ctx       context.Context
+}
+
+const (
+	gzipHeaderSize = 10
+
+	//It is used as a length of a memory region in bytes that is used to buffer
+	//the session recording. 4455 comes from the estimated typical terminal size in
+	//bytes (height=135 width=33) multiplied by 4 bytes of of terminal codes to get
+	//an estimate of a typical screen frame size in bytes.
+	RecorderBufferSize = 4 * 4455
+)
+
+func NewRecorder(ctx context.Context, sessionID string, store store.DataStore) *Recorder {
+	return &Recorder{
+		ctx:       ctx,
+		sessionID: sessionID,
+		store:     store,
+	}
+}
+
+func (r *Recorder) Write(d []byte) (n int, err error) {
+	if len(r.sessionID) < 1 {
+		return 0, nil
+	}
+
+	var buffer bytes.Buffer
+	gzipWriter := gzip.NewWriter(&buffer)
```

Comment:
You get better compression if you preserve the compression context. How about keeping the writer as an attribute in `Recorder`?

Reply:
I completely agree, but I want to be able to decompress each document separately. Imagine a scenario where we build a "player" in the UI or do some debugging, or we want to check what commands where executed in a given time interval.
(also see above comment)